### PR TITLE
add support to youtube playlist and fix thumbnail crop

### PR DESCRIPTION
**Pull request by iamprem**
- Added support to download songs from Youtube playlist
- Thumbnail from youtube usually have left and right padding to fit the aspect ratio. Fixed it to crop out the padded area.
- Updating playlist name was causing it to redownload into new folder. Updated playlist to use UUID to properly de-dupe.

Have oppurtunities to tidy up this code, but wanted to sent this first revision to see if you are open to pull requests.

### CODE REVIEW

**Comment by TheWicklowWolf** [comment]
Yes, I'm open to pull requests, and this seems like a valuable addition. Thank you for contributing!

**A few notes:**
1. This should ideally support playlists from both YouTube and YouTube Music.
2. Regarding the thumbnails, could you provide a screenshot to clarify what you're referring to? As I haven't noticed this issue myself..
3. The playlist name is used as the folder name, so while using a Playlist UUID is a useful idea, I'm not sure if its much of an improvement.
   - The current **correct** flow to change an playlist name is:
     - Hit Edit -> Rename "Playlist Name" 
     - Save changes and Close the Popup
     - Then "Save Sync List" on the main UI
     - Manually rename folder

So I'm not entirely sure if this is worth perusing; if you change the playlist name in the UI, you would just need to manually update the folder name to match. 
Additionally, you would have to address existing playlists that don't have a UUID.

File: src/Syncify.py [hunk 1]
```diff
@@ -224,9 +248,14 @@ def get_download_list(self, playlist):
...
-                    futures.append((future, cleaned_full_file_name))
-                    self.logger.warning("Searching for Song: " + cleaned_full_file_name)
+                    if song.get('VideoID'):
+                        song_list_to_download.append({"title": cleaned_full_file_name,
+                                                      "link": "https://www.youtube.com/watch?v=" + song['VideoID']})
+                        self.logger.warning("Added Song to Download List: " + cleaned_full_file_name + " : " + song['VideoID'])
+                    else:
+                        future = executor.submit(self.find_youtube_link, song_artist, song_title)
+                        futures.append((future, cleaned_full_file_name))
+                        self.logger.warning("Searching for Song: " + cleaned_full_file_name)
```
Comment (TheWicklowWolf) [inline src/Syncify.py:258]: I wonder if this could be moved into the `if "youtube" in playlist_link:` ?

Comment (TheWicklowWolf) [reply src/Syncify.py:258]: There might be a bit of re-arranging though

File: src/Syncify.py [hunk 2]
```diff
@@ -162,6 +162,27 @@ def spotify_extractor(self, link):
...
+            playlist_name = playlist['title']
+
+            for track in playlist['tracks']:
+                track_title = track['title']
+                artist_str = ", ".join([a['name'] for a in track['artists']])
+                track_list.append({"Artist": artist_str, "Title": track_title, "Status": "Queued", "Folder": playlist_name, "VideoID": track['videoId']})
+        else:
+            self.logger.error("Unsupported url")
+
+        return track_list
```
Comment (TheWicklowWolf) [inline src/Syncify.py:183]: It would be great if this could handle YouTube and YouTube music here

File: src/Syncify.py [hunk 3]
```diff
@@ -262,7 +291,8 @@ def download_song(self, song, playlist):
         full_file_path = os.path.join(self.playlist_folder_path, title)
         ydl_opts = {
             "ffmpeg_location": "/usr/bin/ffmpeg",
-            "format": "251/best",
+            "format": "bestaudio[ext=m4a]",
```
Comment (TheWicklowWolf) [inline src/Syncify.py:294]: Do you need a fallback here?

File: src/Syncify.py (hunk 1 above)
Comment (iamprem) [reply src/Syncify.py:258]: Moving this to the top if block is resulting in more code duplication. e.g. `cleaned_full_file_name` construction and checking if it is already in the directory will be duplicated. For now, I'm leaving this as-is. We could potentially refactor this method later.

**Comment by iamprem** [comment]
Thanks for the feedback.

Do you see how it has the padding on the the either side of the album art. The post processor command basically eliminates it by cropping the  longer dimension to match the shorter dimension and makes it square.

Heavydirtysoul - twenty one pilots


 - I'll revert the playlist ID change as I misunderstood the intented use. 
 - I'll also revert the audio format as it seems that 251 is the best quality from youtube. Maybe we can fallback to `bestaudio` instead of `best` because `best` looks for a format that has both audio and video and I'm not sure if it will always have the best audio quality.

**Supported URL formats:**
As long as the url has `youtube` in it and a `list` query params with a valid `playlistId` it should work. Tested on the two format that I could get from YTMusic and YT.
- https://music.youtube.com/playlist?list=PLrs49TmHs0BtBvAnt7bvig5SqGNG7S3hT
- https://www.youtube.com/watch?v=w1Smzzw_w7Q&list=PLrs49TmHs0BtBvAnt7bvig5SqGNG7S3hT

I'll create another clean branch and send a PR from that instead of this since it has some unwanted changes and I'm not sure how github prs work :)

**Comment by iamprem** [comment]
Let me know if this looks good to you.

**Comment by TheWicklowWolf** [comment]
> Thanks for the feedback.
> 
> Do you see how it has the padding on the the either side of the album art. The post processor command basically eliminates it by cropping the longer dimension to match the shorter dimension and makes it square.

So what does it do when the image doesn't have any padding?

>     * I'll revert the playlist ID change as I misunderstood the intented use.
> 
>     * I'll also revert the audio format as it seems that 251 is the best quality from youtube. Maybe we can fallback to `bestaudio` instead of `best` because `best` looks for a format that has both audio and video and I'm not sure if it will always have the best audio quality.

Yes, fallback to `bestaudio` is a good idea.


> **Supported URL formats:** As long as the url has `youtube` in it and a `list` query params with a valid `playlistId` it should work. Tested on the two format that I could get from YTMusic and YT.
> 
>     * https://music.youtube.com/playlist?list=PLrs49TmHs0BtBvAnt7bvig5SqGNG7S3hT
> 
>     * https://www.youtube.com/watch?v=w1Smzzw_w7Q&list=PLrs49TmHs0BtBvAnt7bvig5SqGNG7S3hT

Perfect.

**Comment by TheWicklowWolf** [comment]
> Let me know if this looks good to you.

This looks good, thanks again for you effort. 

Could you also make the cropping feature conditional based on an environment variable, with the default set to off? 
This is mainly because I’ll likely use this in other projects and want to add the feature without altering the default behavior.

**Comment by iamprem** [comment]
To keep all  configurations in one place, should we add "Force square album art" checkbox in the "Configuration" modal instead?

**Comment by TheWicklowWolf** [comment]
> To keep all configurations in one place, should we add "Force square album art" checkbox in the "Configuration" modal instead?

No, the "Configuration" modal is kinda full already. I think just put it as an environmental variable and include the details in the readme.

**Comment by iamprem** [comment]
Moved the album art crop behind an env variable and updated readme. Let me know if this is good.

**Comment by TheWicklowWolf** [comment]
Two minor things:
- I don't think it's necessary to define `ENV crop_album_art=false` in the Dockerfile. While it's probably good practice, I didn't include it in any of the other projects, so could you remove it for consistency?
- Would you like to change `"format": "251/best"` to `"format": "251/bestaudio"`?

I'll merge it then. Sorry for being a bit pedantic, and I appreciate your efforts here.

**Comment by iamprem** [comment]
No worries. Addressed both comments.

**Comment by TheWicklowWolf** [comment]
Thanks again for your efforts on this.

I apologize for the code quality; it was put together somewhat haphazardly without much planning, so it’s a bit rough.

If there's anything else you'd like to work on, feel free to submit pull requests. All the best!

**Pull request merged**